File: apps/api/app/services/stimulus_parser.py
"""Input Parser — first box in the architecture.

Turns raw pasted text (product concept / landing page / ad / pricing table)
into StimulusFeatures: prices, proof signals, risk signals, salient tokens and
a category guess. Both the mock reaction engine and the quality metrics consume
these features, and real LLM providers receive them as structured context, so
this module stays useful after the mock is retired.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from ..utils.text import salient_tokens, tokenize

_PRICE_RE = re.compile(
    r"(?:(?:\$|usd\s?|€|£)\s?(\d{1,6}(?:[.,]\d{1,2})?))|(?:(\d{1,6}(?:[.,]\d{1,2})?)\s?(?:usd|\$|/mo|/month|per month|per user|/user|/seat|per seat))",
    re.IGNORECASE,
)

_CATEGORY_KEYWORDS: dict[str, list[str]] = {
    "saas_b2b": ["saas", "crm", "workflow", "dashboard", "integration", "api",
                 "analytics", "automation", "team", "workspace", "sso", "seat"],
    "consumer_app": ["app", "mobile", "download", "social", "photo", "video",
                     "streak", "friends", "feed", "ios", "android"],
    "ecommerce": ["shipping", "checkout", "cart", "store", "order", "delivery",
                  "marketplace", "returns", "discount"],
    "fintech": ["payment", "wallet", "bank", "invest", "loan", "credit",
                "crypto", "invoice", "payout", "remittance"],
    "edtech": ["course", "learn", "tutor", "exam", "study", "students",
               "lesson", "certification", "ielts", "toeic"],
    "health_fitness": ["fitness", "workout", "health", "sleep", "calories",
                       "meditation", "therapy", "wellness"],
    "devtool": ["developer", "sdk", "cli", "deploy", "code", "repo", "docker",
                "kubernetes", "observability", "gpu", "inference"],
}

_AI_WORDS = {"ai", "gpt", "llm", "ml", "model", "copilot", "agent", "genai",
             "machine", "neural", "automated", "automation"}
_PROOF_WORDS = {"testimonial", "case", "study", "customers", "reviews", "rated",
                "trusted", "results", "guarantee", "refund", "money-back",
                "benchmark", "proven", "roi", "%"}
_TRIAL_WORDS = {"trial", "free", "freemium", "demo", "sandbox"}
_SUBSCRIPTION_WORDS = {"subscription", "monthly", "yearly", "annual", "/mo",
                       "recurring", "plan", "tier", "billed"}
_PRIVACY_WORDS = {"privacy", "data", "gdpr", "encrypted", "encryption",
                  "tracking", "personal", "anonymous"}
_SECURITY_WORDS = {"soc2", "soc-2", "iso27001", "sso", "saml", "compliance",
                   "audit", "security", "hipaa"}
_LOCKIN_WORDS = {"contract", "commitment", "lock-in", "annual", "cancel",
                 "minimum"}


@dataclass
class StimulusFeatures:
    raw: str
    title: str
    stimulus_type: str
    tokens: list[str] = field(default_factory=list)
    anchors: list[str] = field(default_factory=list)  # salient product tokens
    prices: list[float] = field(default_factory=list)
    min_price: float | None = None
    max_price: float | None = None
    has_pricing: bool = False
    has_free_trial: bool = False
    has_proof: bool = False
    has_guarantee: bool = False
    mentions_ai: bool = False
    mentions_subscription: bool = False
    mentions_privacy: bool = False
    mentions_security: bool = False
    mentions_lockin: bool = False
    jargon_score: float = 0.0   # 0 = plain language, 1 = buzzword soup
    clarity_score: float = 0.5  # crude readability proxy
    category: str = "other"

    def anchor_set(self) -> set[str]:
        return set(self.anchors)


def parse_stimulus(raw: str, title: str, stimulus_type: str) -> StimulusFeatures:
    text = raw.strip()
    toks = tokenize(text)
    tokset = set(toks)
    f = StimulusFeatures(raw=text, title=title, stimulus_type=stimulus_type)
    f.tokens = toks
    f.anchors = salient_tokens(text)
    # Title words are anchors too — reactions naming the product count as grounded.
    f.anchors.extend(t for t in tokenize(title) if len(t) > 2 and t not in f.anchors)

    # --- prices ---------------------------------------------------------------
    for m in _PRICE_RE.finditer(text):
        val = m.group(1) or m.group(2)
        if val:
            try:
                price = float(val.replace(",", ""))
                if 0 < price < 1_000_000:
                    f.prices.append(price)
            except ValueError:
                pass
    if f.prices:
        f.has_pricing = True
        f.min_price = min(f.prices)
        f.max_price = max(f.prices)

    # --- signal flags -----------------------------------------------------------
    lower = text.lower()
    f.has_free_trial = bool(tokset & _TRIAL_WORDS)
    # Proof = proof vocabulary, percentage claims, or numeric social proof
    # ("1,200 stores", "used by 4,000 families") — skeptical personas key on this.
    f.has_proof = (
        bool(tokset & _PROOF_WORDS)
        or bool(re.search(r"\d{2,}\s?[%+]", text))
        or bool(re.search(
            r"\b\d[\d,.]*k?\s+(?:users|customers|teams|families|stores|companies|"
            r"businesses|agencies|schools|developers|sellers|clients|members)\b",
            text, re.IGNORECASE))
    )
    f.has_guarantee = "guarantee" in lower or "money-back" in lower or "refund" in lower
    f.mentions_ai = bool(tokset & _AI_WORDS)
    f.mentions_subscription = bool(tokset & _SUBSCRIPTION_WORDS) or "/mo" in lower
    f.mentions_privacy = bool(tokset & _PRIVACY_WORDS)
    f.mentions_security = bool(tokset & _SECURITY_WORDS)
    f.mentions_lockin = bool(tokset & _LOCKIN_WORDS)

    # --- jargon & clarity -------------------------------------------------------
    if toks:
        buzz = sum(1 for t in toks if t in _AI_WORDS or len(t) >= 12)
        f.jargon_score = min(1.0, buzz / max(20, len(toks)) * 6)
    sentences = [s for s in re.split(r"[.!?\n]+", text) if s.strip()]
    if sentences:
        avg_len = sum(len(tokenize(s)) for s in sentences) / len(sentences)
        # ~12-18 words/sentence reads well; long sentences hurt clarity
        f.clarity_score = max(0.0, min(1.0, 1.25 - abs(avg_len - 15) / 25))

    # --- category guess -----------------------------------------------------------
    best_cat, best_hits = "other", 0
    for cat, words in _CATEGORY_KEYWORDS.items():
        hits = sum(1 for w in words if w in tokset)
        if hits > best_hits:
            best_cat, best_hits = cat, hits
    if best_hits >= 2:
        f.category = best_cat
    return f

File: apps/web/lib/server/engine/stimulusParser.ts
/**
 * Input Parser — port of apps/api/app/services/stimulus_parser.py.
 * Turns raw pasted text into StimulusFeatures (prices, proof/risk signals,
 * salient tokens, category guess) consumed by the reaction engine + quality.
 */

import { salientTokens, tokenize } from "./text";

// Global regex; we iterate matches manually.
const PRICE_RE =
  /(?:(?:\$|usd\s?|€|£)\s?(\d{1,6}(?:[.,]\d{1,2})?))|(?:(\d{1,6}(?:[.,]\d{1,2})?)\s?(?:usd|\$|\/mo|\/month|per month|per user|\/user|\/seat|per seat))/gi;

const CATEGORY_KEYWORDS: Record<string, string[]> = {
  saas_b2b: ["saas", "crm", "workflow", "dashboard", "integration", "api", "analytics", "automation", "team", "workspace", "sso", "seat"],
  consumer_app: ["app", "mobile", "download", "social", "photo", "video", "streak", "friends", "feed", "ios", "android"],
  ecommerce: ["shipping", "checkout", "cart", "store", "order", "delivery", "marketplace", "returns", "discount"],
  fintech: ["payment", "wallet", "bank", "invest", "loan", "credit", "crypto", "invoice", "payout", "remittance"],
  edtech: ["course", "learn", "tutor", "exam", "study", "students", "lesson", "certification", "ielts", "toeic"],
  health_fitness: ["fitness", "workout", "health", "sleep", "calories", "meditation", "therapy", "wellness"],
  devtool: ["developer", "sdk", "cli", "deploy", "code", "repo", "docker", "kubernetes", "observability", "gpu", "inference"],
};

const AI_WORDS = new Set(["ai", "gpt", "llm", "ml", "model", "copilot", "agent", "genai", "machine", "neural", "automated", "automation"]);
const PROOF_WORDS = new Set(["testimonial", "case", "study", "customers", "reviews", "rated", "trusted", "results", "guarantee", "refund", "money-back", "benchmark", "proven", "roi", "%"]);
const TRIAL_WORDS = new Set(["trial", "free", "freemium", "demo", "sandbox"]);
const SUBSCRIPTION_WORDS = new Set(["subscription", "monthly", "yearly", "annual", "/mo", "recurring", "plan", "tier", "billed"]);
const PRIVACY_WORDS = new Set(["privacy", "data", "gdpr", "encrypted", "encryption", "tracking", "personal", "anonymous"]);
const SECURITY_WORDS = new Set(["soc2", "soc-2", "iso27001", "sso", "saml", "compliance", "audit", "security", "hipaa"]);
const LOCKIN_WORDS = new Set(["contract", "commitment", "lock-in", "annual", "cancel", "minimum"]);

export interface StimulusFeatures {
  raw: string;
  title: string;
  stimulusType: string;
  tokens: string[];
  anchors: string[];
  prices: number[];
  minPrice: number | null;
  maxPrice: number | null;
  hasPricing: boolean;
  hasFreeTrial: boolean;
  hasProof: boolean;
  hasGuarantee: boolean;
  mentionsAi: boolean;
  mentionsSubscription: boolean;
  mentionsPrivacy: boolean;
  mentionsSecurity: boolean;
  mentionsLockin: boolean;
  jargonScore: number;
  clarityScore: number;
  category: string;
}

export function anchorSet(f: StimulusFeatures): Set<string> {
  return new Set(f.anchors);
}

function intersects(a: Set<string>, b: ReadonlySet<string>): boolean {
  for (const x of a) if (b.has(x)) return true;
  return false;
}

export function parseStimulus(raw: string, title: string, stimulusType: string): StimulusFeatures {
  const text = raw.trim();
  const toks = tokenize(text);
  const tokset = new Set(toks);

  const anchors = salientTokens(text);
  for (const t of tokenize(title)) {
    if (t.length > 2 && !anchors.includes(t)) anchors.push(t);
  }

  const prices: number[] = [];
  let m: RegExpExecArray | null;
  PRICE_RE.lastIndex = 0;
  while ((m = PRICE_RE.exec(text)) !== null) {
    const val = m[1] || m[2];
    if (val) {
      const price = parseFloat(val.replace(/,/g, ""));
      if (Number.isFinite(price) && price > 0 && price < 1_000_000) prices.push(price);
    }
  }
  const hasPricing = prices.length > 0;
  const minPrice = hasPricing ? Math.min(...prices) : null;
  const maxPrice = hasPricing ? Math.max(...prices) : null;

  const lower = text.toLowerCase();
  const hasFreeTrial = intersects(tokset, TRIAL_WORDS);
  const hasProof =
    intersects(tokset, PROOF_WORDS) ||
    /\d{2,}\s?[%+]/.test(text) ||
    new RegExp(
      "\\b\\d[\\d,.]*k?\\s+(?:users|customers|teams|families|stores|companies|businesses|agencies|schools|developers|sellers|clients|members)\\b",
      "i",
    ).test(text);
  const hasGuarantee = lower.includes("guarantee") || lower.includes("money-back") || lower.includes("refund");
  const mentionsAi = intersects(tokset, AI_WORDS);
  const mentionsSubscription = intersects(tokset, SUBSCRIPTION_WORDS) || lower.includes("/mo");
  const mentionsPrivacy = intersects(tokset, PRIVACY_WORDS);
  const mentionsSecurity = intersects(tokset, SECURITY_WORDS);
  const mentionsLockin = intersects(tokset, LOCKIN_WORDS);

  let jargonScore = 0.0;
  if (toks.length > 0) {
    const buzz = toks.filter((t) => AI_WORDS.has(t) || t.length >= 12).length;
    jargonScore = Math.min(1.0, (buzz / Math.max(20, toks.length)) * 6);
  }
  let clarityScore = 0.5;
  const sentences = text.split(/[.!?\n]+/).filter((s) => s.trim());
  if (sentences.length > 0) {
    const avgLen = sentences.reduce((s, sent) => s + tokenize(sent).length, 0) / sentences.length;
    clarityScore = Math.max(0.0, Math.min(1.0, 1.25 - Math.abs(avgLen - 15) / 25));
  }

  let bestCat = "other";
  let bestHits = 0;
  for (const [cat, words] of Object.entries(CATEGORY_KEYWORDS)) {
    const hits = words.filter((w) => tokset.has(w)).length;
    if (hits > bestHits) {
      bestCat = cat;
      bestHits = hits;
    }
  }
  const category = bestHits >= 2 ? bestCat : "other";

  return {
    raw: text,
    title,
    stimulusType,
    tokens: toks,
    anchors,
    prices,
    minPrice,
    maxPrice,
    hasPricing,
    hasFreeTrial,
    hasProof,
    hasGuarantee,
    mentionsAi,
    mentionsSubscription,
    mentionsPrivacy,
    mentionsSecurity,
    mentionsLockin,
    jargonScore,
    clarityScore,
    category,
  };
}
